source/Irrlicht/CIrrDeviceSDL.cpp

```diff
@@ -105,6 +107,92 @@ EM_BOOL CIrrDeviceSDL::MouseLeaveCallback(int eventType, const EmscriptenMouseEv
 }
 #endif
 
+
+bool CIrrDeviceSDL::keyIsKnownSpecial(EKEY_CODE key)
+{
+	switch ( key )
+	{
+		// keys which are known to have safe special character interpretation
+		// could need changes over time (removals and additions!)
+		case KEY_RETURN:
+		case KEY_PAUSE:
+		case KEY_ESCAPE:
+		case KEY_PRIOR:
+		case KEY_NEXT:
+		case KEY_HOME:
+		case KEY_END:
+		case KEY_LEFT:
+		case KEY_UP:
+		case KEY_RIGHT:
+		case KEY_DOWN:
+		case KEY_TAB:
+		case KEY_PRINT:
+		case KEY_SNAPSHOT:
+		case KEY_INSERT:
+		case KEY_BACK:
+		case KEY_DELETE:
+		case KEY_HELP:
+		case KEY_APPS:
+		case KEY_SLEEP:
+		case KEY_F1:
+		case KEY_F2:
+		case KEY_F3:
+		case KEY_F4:
+		case KEY_F5:
+		case KEY_F6:
+		case KEY_F7:
+		case KEY_F8:
+		case KEY_F9:
+		case KEY_F10:
+		case KEY_F11:
+		case KEY_F12:
+		case KEY_F13:
+		case KEY_F14:
+		case KEY_F15:
+		case KEY_F16:
+		case KEY_F17:
+		case KEY_F18:
+		case KEY_F19:
+		case KEY_F20:
+		case KEY_F21:
+		case KEY_F22:
+		case KEY_F23:
+		case KEY_F24:
+		case KEY_NUMLOCK:
+		case KEY_SCROLL:
+		case KEY_LCONTROL:
+		case KEY_RCONTROL:
+			return true;
+
+		default:
+			return false;
+	}
+}
+
+int CIrrDeviceSDL::findCharToPassToIrrlicht(int assumedChar, EKEY_CODE key) {
+	switch (key) {
+		case KEY_PRIOR:
+		case KEY_NEXT:
+		case KEY_HOME:
+		case KEY_END:
+		case KEY_LEFT:
+		case KEY_UP:
+		case KEY_RIGHT:
+		case KEY_DOWN:
+			return 0;
+		default:
+			return assumedChar;
+	}
+}

```

Comment:
Pressing num-lock in an edit-box produces some sort of horizontal tab for me.
This fixes it:
```cpp
	// see https://wiki.libsdl.org/SDL2/SDLKeycodeLookup
	if (assumedChar & (1<<30))
		return 0;
```

Idk if the other cases still need to return 0 because "These keys are handled differently in CGUIEditBox.cpp".
(But it doesn't seem so, according to my testing.)

(Note: The sdl keycode is a unicode char if it's printable. I'm not sure if it's appropriate to convert it to a `wchar_t`.)

Reply:
I've added numlock as a special case, let me know if that works.

Reply:
Works.
But I'd still recommend to add the check I gave above, as the keycode is not a character if that bit is set.
To emphasize what the wiki says (see link):
>Keycodes for keys with printable characters are represented by the character byte in parentheses. Keycodes without character representations are determined by their [scancode](https://wiki.libsdl.org/SDL2/SDL_Scancode) bitwise OR-ed with 1<<30 (0x40000000).

Also (see <https://github.com/libsdl-org/sdl2-compat/blob/bb66ed056b8a803677b23082438677a5676e3f22/include/SDL2/SDL_keycode.h#L37-L40>):
> Values of this type are used to represent keyboard keys using the current
> layout of the keyboard.  These values include Unicode values representing
> the unmodified character that would be generated by pressing the key, or
> an SDLK_* constant for those keys that do not generate characters.

Reply:
That's a really nice way to check, thanks!